tests/std/tests/P0896R4_counted_iterator/test.cpp

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <concepts>
+#include <iterator>
+
+#include <range_algorithm_support.hpp>
+using namespace std;
+
+struct instantiator {
+    template <input_or_output_iterator Iter>
+    static constexpr void call() {
+        if constexpr (is_permissive) { // TRANSITION, FIXME (File an issue to investigate this later?)
+            (void) input_iterator<Iter>;
+        }
+
+        int input[5] = {1, 2, 3, 4, 5};
+        // [counted.iter.const]
+        {
+            [[maybe_unused]] counted_iterator<Iter> defaultConstructed{};
+
+            [[maybe_unused]] counted_iterator<Iter> constructed(Iter{input}, iter_difference_t<Iter>{2});
+            [[maybe_unused]] counted_iterator<Iter> constructedEmpty{Iter{input}, iter_difference_t<Iter>{0}};
+
+            //[[maybe_unused]] counted_iterator<Iter> convertibleConstructed{};
+            //[[maybe_unused]] counted_iterator<Iter> convertibleAssigned{};
+        }
+
+        // [counted.iter.access]
+        {
+            if constexpr (copy_constructible<Iter>) {
+                const counted_iterator<Iter> lvalueBase{Iter{input}, 2};
+                const same_as<Iter> auto base1 = lvalueBase.base();
+                assert(base1.peek() == Iter{input}.peek());
+            }
+
+            const same_as<Iter> auto base2 = counted_iterator<Iter>{Iter{input}, 2}.base();
+            assert(base2.peek() == Iter{input}.peek());
+
+            const same_as<iter_difference_t<Iter>> auto length = counted_iterator<Iter>{Iter{input}, 2}.count();
+            assert(length == 2);
+        }
+
+        // [counted.iter.elem]
+        {
+            auto iter = counted_iterator<Iter>{Iter{input}, 2};
+            assert(*iter == 1);
+
+            const auto constIter = counted_iterator<Iter>{Iter{input}, 2};
+            assert(*constIter == 1);
+
+            if constexpr (random_access_iterator<Iter>) {

```

Comment:
Pure style: It may be simpler to organize the test cases by requirement rather than by subclause.

Reply:
That might be true but so it is easier to verify correctness and matching requirements.

Not a hill to die on though

Reply:
Fair enough: like I said, this is pure style so you're free to ignore. I have the habit of ordering everything by requirement from the range-v3 days when it made it easier for us to validate things like "this type correctly models `bidirectional_iterator`" by reducing to "everything up to here correctly models `forward_iterator`, and "here are implementations of `operator--()` and `operator--(int)` that meet the semantic requirements for bidits." Now that there's a spec to validate each of the bits against, it makes sense for the order to follow the spec order instead.